dataflow: correctly sync tables added to Postgres publication

JLDLaughlin opened the pull request:
Fixes https://github.com/MaterializeInc/materialize/issues/6742.

Our Postgres source ingest data from upstream [publications](https://www.postgresql.org/docs/13/logical-replication-publication.html). Currently, we don't correctly ingest data from upstream relations if they are added to a publication _after_ a Postgres source has been made. This is because we simply start ingesting `INSERT`, `UPDATE`, and `DELETE` messages without taking an initial snapshot of the data in the newly added relation.

This PR changes that by handling incoming `Relation` messages. During logical replication, [`Relation` messages are sent](https://www.postgresql.org/docs/13/protocol-logical-replication.html#PROTOCOL-LOGICAL-MESSAGES-FLOW) when 1) it's the first time we're seeing a DML message for a relation in this session, or 2) because the relation definition has changed. This PR handles the first of those two cases in order to ingest the added relations correctly.

Interesting behavior to note: `Relation` messages aren't sent unless there is some sort of a data event for that relation (`INSERT`, `UPDATE`, or `DELETE`). This means that we aren't actually able to snapshot a relation immediately when it's added to a publication. This makes the row handling a bit more complicated, and is captured in the `add-relation-to-publication.td` test.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/materializeinc/materialize/6927)
<!-- Reviewable:end -->

JLDLaughlin commented on src/dataflow/src/source/postgres.rs:
```diff
@@ -197,22 +190,69 @@ impl PostgresSourceReader {
+    /// Takes a snapshot of the specified relation, syncing the data to the source.
+    /// This function will be called when we detect a new relation has been added to the
+    /// Postgres publication.
+    async fn produce_snapshot(&mut self, rel_id: u32) -> Result<Vec<Row>, ReplicationError> {
```
There is definitely redundancy between this function and `produce_initial_snapshot`. I made them separate to avoid putting a bunch of `if/elses` in.

philip-stoev commented on test/pg-cdc-resumption/verify-data.td:
```diff
@@ -7,7 +7,11 @@
 > SELECT COUNT(*) = 500000 FROM t1;
 true
 > SELECT COUNT(*) = 500000 FROM t2;
 true
```
A COUNT against t3 is needed here.

philip-stoev commented on test/pg-cdc-resumption/add-relation-to-publication.td:
```diff
@@ -0,0 +1,13 @@
+# by the Apache License, Version 2.0.
+
+$ postgres-execute connection=postgres://postgres:postgres@postgres
+ALTER PUBLICATION mz_source ADD TABLE t3;
```
I think for a most-stressful test, t3 should contain data prior to adding it to the publication, if that scenario is supported.

In this case, the INSERT below can be replaced with a DELETE, in the pattern of the ```delete*.td``` tests.

JLDLaughlin replied:
Done! Thank you for the suggestion!

philip-stoev commented on test/pg-cdc-resumption/mzcompose.yml:
```diff
@@ -191,15 +213,15 @@ services:
     command: --experimental --disable-telemetry
     ports:
-      - 6875
+      - 6875:6875
```
I think host port selection should be left dynamic. Otherwise people that have materialize or postgres running locally will see conflicts.

philip-stoev commented on test/pg-cdc-resumption/mzcompose.yml:
```diff
@@ -191,15 +213,15 @@ services:
     image: postgres:11.4
     ports:
-      - 5432
+      - 5432:5432
```
ditto here

philip-stoev commented on test/pg-cdc/add-relation-to-publication.td:
```diff
@@ -0,0 +1,42 @@
+INSERT INTO nonpk_table VALUES (3, 3);
+
+> SELECT COUNT(*) FROM mz_source
+7
```
Please also SELECT from nonpk_table here .

philip-stoev commented on src/dataflow/src/source/postgres.rs:
```diff
@@ -232,6 +272,71 @@ impl PostgresSourceReader {
+        slot_name: &str,
+    ) -> Result<(), ReplicationError> {
+        let _ = client
+            .simple_query(&format!("DROP_REPLICATION_SLOT {:?}", slot_name))
```
Is ```WAIT``` needed here?

philip-stoev reviewed (approved):
Thank you very very much for the tests! It is exactly what I would have added. Much appreciated.

petrosagg reviewed (commented):
I haven't done a detailed pass over the code yet but I think there is a more general issue that needs to be addressed first. When we are consuming a replication stream and a relation is added to the publication after the fact the current code will snapshot it at some LSN `snapshot_lsn`, but then will continue streaming updates starting from `self.lsn`.

This will cause all updates to the new relation that hapened between `snapshot_lsn` and `self.lsn` to disappear, or, if `self.lsn` is behind `snapshot_lsn` to be reported doubly.

petrosagg commented on src/dataflow/src/source/postgres.rs:
```diff
@@ -197,22 +189,76 @@ impl PostgresSourceReader {
+            .await?;
+
+        let lsn: PgLsn = client
+            .query_one("SELECT pg_current_wal_lsn()", &[])
```
This function unfortunately doesn't return the lsn of this transaction's snapshot, it just tells you what the latest LSN of the server is. I think we need to get the consistent point from the side effect of creating a slot and then immediately discard it after we're done with the transaction

petrosagg commented on src/dataflow/src/source/postgres.rs:
```diff
@@ -197,22 +189,76 @@ impl PostgresSourceReader {
+    async fn produce_snapshot(
+        &mut self,
+        rel_id: u32,
+    ) -> Result<(PgLsn, Vec<Row>), ReplicationError> {
```
This could be a potentially large table, could we reuse the snapshot mechanism we use for the initial snapshot so that we don't store everything in a Vec?

JLDLaughlin commented:
> Thank you very very much for the tests! It is exactly what I would have added. Much appreciated.

@philip-stoev this is truly the kindest compliment!

@petrosagg--I'm ready for you to take another look! Just want to point out some unfortunately quirky behavior first:

When a user adds a relation to publication, nothing happens. We only get the `Relation` message as an `Insert`, `Update`, or `Delete` message for that relation is sent. Once we receive the `Relation` and data message, and we take a `SELECT *` snapshot of the upstream relation, the `consistent_point` of that relation will be ahead of the `Insert`, `Update`, or `Delete` that was sent. This means after a user adds a relation to the source publication, we will need to see at least two data events in order for the snapshotted rows to be inserted into the source.

This is weird! I couldn't figure out a consistent way around it, but a million percent open to suggestions if I'm missing something there.

petrosagg reviewed (commented):
I think it looks good but the code duplication makes me a bit uncomfortable. The `produce_snapshot` and `produce_initial_snapshot` functions are very similar, same for the two different behaviours of `publication_info`.

It's worth trying to express both the initial snapshot phase and a subsequent re-sync phase as the same thing.

For example a function that "takes a snapshot for all tables in the publication that have been synced yet" would provide a unified implementation for both cases and we can just call that whenever we are initializing the source and whenever we are encountering a Relation message. Requesting the full list of tables and checking them against the HashSet is a small performance price to pay for less code to maintain. If we had this function, we could also be running it periodically to proactively add tables.

> This means after a user adds a relation to the source publication, we will need to see at least two data events in order for the snapshotted rows to be inserted into the source.

I'm not sure I follow, maybe we need to discuss this over zoom for high bandwidth. My understanding is that once we see the first data event for a new table we'll commit a snapshot of that table without requiring a second event

JLDLaughlin commented:
I just added another commit to illustrate the problem that I highlighted in an earlier comment, and that Petros and I discussed in a Zoom call last Friday. The problem is that the `consistent_point` of the queried data from the newly added relation is always ahead of our replication stream's LSN. If the publication isn't producing a lot of data events, the replication stream will never reach that `consistent_point` and, therefore, will never add the new relation to the source.

Here is a sample run illustrating the issue. I ran the updated `pg-cdc/add-relation-to-replication.td` test with some additional log lines (that can be found in the newest commit):
```
materialized v0.8.1-dev (8e00cefac) listening on 0.0.0.0:6875...
=======================================================================
Thank you for trying Materialize!

We are interested in any and all feedback you have, which may be able
to improve both our software and your queries! Please reach out at:

    Web: https://materialize.com
    GitHub issues: https://github.com/MaterializeInc/materialize/issues
    Email: support@materialize.io
    Twitter: @MaterializeInc
=======================================================================

!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
                                WARNING!
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
Starting Materialize in experimental mode means:

- This node's catalog of views and sources are unstable.

If you use any version of Materialize besides this one, you might
not be able to start the Materialize node. To fix this, you'll have
to remove all of Materialize's data (e.g. rm -rf mzdata) and start
the node anew.

- You must always start this node in experimental mode; it can no
longer be started in non-experimental/regular mode.

For more details, see https://materialize.com/docs/cli#experimental-mode
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!

2021-06-22T04:06:04.921157Z  INFO coord::catalog: create database materialize
2021-06-22T04:06:04.921194Z  INFO coord::catalog: create schema materialize.public
2021-06-22T04:06:04.988952Z  INFO coord::catalog: create source materialize.public.mz_source (u1)
2021-06-22T04:06:04.989049Z  INFO coord::catalog: create index materialize.public.mz_source_primary_idx (u2)
2021-06-22T04:06:04.996582Z  INFO coord::timestamp: Timestamping Source u1 with Real Time Consistency.
2021-06-22T04:06:05.022107Z  INFO dataflow::source::postgres: source materialize.public.mz_source: created Postgres replication slot materialize_e1dc74400af3469ba1d045f752a71c18
2021-06-22T04:06:05.022150Z  INFO dataflow::source::postgres: source materialize.public.mz_source: consistent LSN for materialize_e1dc74400af3469ba1d045f752a71c18 is 0/1676340
2021-06-22T04:06:05.023109Z  INFO dataflow::source::postgres: source materialize.public.mz_source: replication snapshot succeeded
2021-06-22T04:06:06.138823Z  INFO dataflow::source::postgres: source materialize.public.mz_source: created Postgres replication slot materialize_e1dc74400af3469ba1d045f752a71c18_664c0dcce7d54ba2b0
2021-06-22T04:06:06.138871Z  INFO dataflow::source::postgres: source materialize.public.mz_source: consistent LSN for materialize_e1dc74400af3469ba1d045f752a71c18_664c0dcce7d54ba2b0 is 0/16767A8
2021-06-22T04:06:06.141564Z  INFO dataflow::source::postgres: source materialize.public.mz_source: dropped Postgres replication slot materialize_e1dc74400af3469ba1d045f752a71c18_664c0dcce7d54ba2b0
2021-06-22T04:06:06.141910Z  INFO dataflow::source::postgres: source materialize.public.mz_source: self.lsn is 0/1676700, got XLogData wal_start 0/1676700, wal_end 0/1676700
2021-06-22T04:06:06.141953Z  INFO dataflow::source::postgres: source materialize.public.mz_source: self.lsn is 0/1676700, got XLogData wal_start 0/1676770, wal_end 0/1676770
2021-06-22T04:06:06.141963Z  INFO dataflow::source::postgres: source materialize.public.mz_source: added relations not synced at LSN 0/1676770, waiting until LSN 0/16767A8 to maintain consistency: [16394]
2021-06-22T04:07:05.031502Z  WARN dataflow::source::postgres: replication for source materialize.public.mz_source interrupted, retrying: connection closed
2021-06-22T04:07:08.033847Z  INFO dataflow::source::postgres: resuming replication for source materialize.public.mz_source
2021-06-22T04:07:08.043994Z  INFO dataflow::source::postgres: source materialize.public.mz_source: self.lsn is 0/1676770, got PrimaryKeepAlive wal_end 0/1676340
2021-06-22T04:08:08.046149Z  WARN dataflow::source::postgres: replication for source materialize.public.mz_source interrupted, retrying: connection closed
2021-06-22T04:08:11.048080Z  INFO dataflow::source::postgres: resuming replication for source materialize.public.mz_source
2021-06-22T04:08:11.056214Z  INFO dataflow::source::postgres: source materialize.public.mz_source: self.lsn is 0/1676770, got PrimaryKeepAlive wal_end 0/1676340
2021-06-22T04:09:11.057507Z  WARN dataflow::source::postgres: replication for source materialize.public.mz_source interrupted, retrying: connection closed
2021-06-22T04:09:14.025098Z  INFO dataflow::source::postgres: resuming replication for source materialize.public.mz_source
2021-06-22T04:09:14.036799Z  INFO dataflow::source::postgres: source materialize.public.mz_source: self.lsn is 0/1676770, got PrimaryKeepAlive wal_end 0/1676340
```

As discussed with @petrosagg , I tried comparing the commit's `commit_lsn` instead of `end_lsn`, but that didn't make a difference.

I figured out a way to fix this issue, but unfortunately the functions required to do this weren't added to Postgres until version 12. In Postgres versions 12+, we could create a copy of the streaming replication slot via [`pg_copy_logical_replication_slot`](https://www.postgresql.org/docs/13/functions-admin.html), check if there are any pending changes between `self.lsn` and the max pending LSN via `pg_logical_slot_peek_binary_changes`, and then use `pg_replication_slot_advance` to advance the slot if we're just hanging. I couldn't find a way to create or start a replication slot at a historic LSN before Postgres 12.

Given that the behavior here relies on the characteristics of the underlying data, I don't think it should be merged. I can keep investigating in the meantime!

JLDLaughlin commented:
Also, for posterity, I'm not totally sure _why_ the new consistent LSN can end up beyond the replication stream's LSN! But I did find that the `consistent_point` is assigned by [`SnapBuildFindSnapshot`](https://github.com/postgres/postgres/blob/def5b065ff22a16a80084587613599fe15627213/src/backend/replication/logical/snapbuild.c#L1235), which uses the [LSN of the start of the last record read](https://github.com/postgres/postgres/blob/c2dc19342e05e081dc13b296787baa38352681ef/src/include/access/xlogreader.h#L175).

JLDLaughlin commented:
So I realized that I should have checked the WAL to see what exactly was going on here. I decided to do that by simplifying the code: I created a new branch (without these changes) that creates a new replication slot each time we see a `Relation` message in order to log its consistent point. Then, I ran the `add-relation-to-publication.td` test manually.

Like before, I found that the consistent point of the newly created slot is beyond an LSN that can be reached by the original slot:
```
...
2021-07-14T22:31:10.097650Z  INFO dataflow::source::postgres: BEGIN final LSN is 0/808EE80
2021-07-14T22:31:10.107886Z  INFO dataflow::source::postgres: got relation messsage, consistent point LSN is 0/808EEE8
2021-07-14T22:31:10.108081Z  INFO dataflow::source::postgres: COMMIT LSN is 0/808EE80, end LSN is 0/808EEB0
```

Looking at the WAL, I see that the entries between `0/808EE80` and `0/808EEE8` are `Standby` messages. 
```
rmgr: Transaction len (rec/tot):     46/    46, tx:       3586, lsn: 0/0808EE80, prev 0/0808ED70, desc: COMMIT 2021-07-14 15:31:10.097116 PDT
rmgr: Standby     len (rec/tot):     50/    50, tx:          0, lsn: 0/0808EEB0, prev 0/0808EE80, desc: RUNNING_XACTS nextXid 3587 latestCompletedXid 3586 oldestRunningXid 3587
rmgr: Standby     len (rec/tot):     50/    50, tx:          0, lsn: 0/0808EEE8, prev 0/0808EEB0, desc: RUNNING_XACTS nextXid 3587 latestCompletedXid 3586 oldestRunningXid 3587
```

So we're unable to make progress because we don't advance the LSN when we receive `Standby` messages. We'd need additional `BEGIN`/`INSERT`/`UPDATE`/`DELETE`/`COMMIT` messages to progress past `0/808EEE8`.

@petrosagg -- given this information, would you be open to me pushing a version of this code that uses two slots (as proposed above)? Or, should we just say that relations added to a publication will not be synced (and I can make that change)?

petrosagg commented:
interesting. the standby messages carry the end lsn in them but I'm not sure how to use them in this case. Can you elaborate more on the two slots idea? What data would we publish in the source for the initial sync and then beyond that?

JLDLaughlin commented:
So if each Postgres source had two slots, we could use them as follows:
1. The main replication slot that works exactly as the current slot works.
2. A secondary replication slot that is just used to check if we can progress the LSN of the main replication slot. When we get a standby message, we could call `pg_logical_slot_peek_binary_changes` on this secondary slot to check if there are any changes between the Postgres source's LSN and the standby LSN. If there aren't, advance the main replication slot to the standby LSN. 

This works for me locally! But comes with the extra cost of a replication slot. We would also still need to spin up temporary replication slots to sync relations that are added to a publication.

petrosagg commented:
Interesting! Why can't we do exactly what you describe but on the main slot when we receive a standby message? I need to put a bit more thought on whether there is any case it can go wrong but at the very least it sounds doable with a single slot. Am I missing something?

JLDLaughlin commented:
`pg_logical_slot_peek_binary_changes` can only be called on inactive slots. So, we could use a single slot if we're willing to stop the replication stream on the main slot at some cadence!

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/MaterializeInc/materialize?pullRequest=6927) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/MaterializeInc/materialize?pullRequest=6927) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/MaterializeInc/materialize?pullRequest=6927) it.</sub>

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/MaterializeInc/materialize?pullRequest=6927) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/MaterializeInc/materialize?pullRequest=6927) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/MaterializeInc/materialize?pullRequest=6927) it.</sub>

benesch commented:
Closing as stale. We've learned to support this in the meantime with `ADD SUBSOURCE`.
